File: src/main/java/com/protryon/jasm/Constant.java
package com.protryon.jasm;

public final class Constant<T> {

    public final T value;

    public Constant(T value) {
        this.value = value;
    }

    public String toString() {
        if (value instanceof JType) {
            return value.toString();
        } else if (value instanceof Method) {
            return ((Method) value).descriptor.niceString(((Method) value).name);
        } else if (value instanceof Field) {
            return ((Field) value).name + " " + ((Field) value).type.niceName;
        } else if (value instanceof Local) {
            return "v" + ((Local) value).index;
        }
        return value.toString();
    }

}

File: src/main/java/com/protryon/jasm/Constant.kt
package com.protryon.jasm

class Constant<T>(val value: T) {

    override fun toString(): String {
        if (value is JType) {
            return value.toString()
        } else if (value is Method) {
            return (value as Method).descriptor.niceString((value as Method).name)
        } else if (value is Field) {
            return (value as Field).name + " " + (value as Field).type.niceName
        } else if (value is Local) {
            return "v" + (value as Local).index
        }
        return value.toString()
    }

}
